linera-execution/src/runtime.rs

```diff
@@ -953,64 +972,129 @@ impl ContractSyncRuntime {
-            id: application_id,
-            parameters: description.parameters,
-            signer,
-        });
-        let mut runtime = ContractSyncRuntime::new(runtime);
-        let execution_result = {
-            let mut code = code.instantiate(runtime.clone())?;
-            match action {
-                UserAction::Initialize(context, argument) => code.initialize(context, argument),
-                UserAction::Operation(context, operation) => {
-                    code.execute_operation(context, operation)
-                }
-                UserAction::Message(context, message) => code.execute_message(context, message),
-            }
+        ));
+        let finalize_context = FinalizeContext {
+            authenticated_signer: signer,
+            chain_id,
+            height,
+            next_message_index,
         };
-        // Ensure the `loaded_applications` are cleared to prevent circular references in the
-        // `runtime`
-        runtime.inner().loaded_applications.clear();
-        let mut runtime = runtime
+        runtime.execute(application_id, signer, move |code| match action {
+            UserAction::Initialize(context, argument) => code.initialize(context, argument),
+            UserAction::Operation(context, operation) => code.execute_operation(context, operation),
+            UserAction::Message(context, message) => code.execute_message(context, message),
+        })?;
+        runtime.finalize(finalize_context)?;
+        let runtime = runtime
             .into_inner()
             .expect("Runtime clones should have been freed by now");
-        let execution_outcome = execution_result?;
-        assert_eq!(runtime.call_stack.len(), 1);
-        assert_eq!(runtime.call_stack[0].id, application_id);
-        assert_eq!(runtime.active_applications.len(), 1);
-        assert!(runtime.active_applications.contains(&application_id));
         // Check that all sessions were properly closed.
         if let Some(session_id) = runtime.session_manager.states.keys().next() {
             return Err(ExecutionError::SessionWasNotClosed(*session_id));
         }
-        // Charge for the message grants and add the results of the last call to the
-        // execution results.
-        let outcome = execution_outcome
-            .with_authenticated_signer(signer)
-            .with_refund_grant_to(refund_grant_to)
-            .into_priced(&runtime.resource_controller.policy)?;
-        for message in &outcome.messages {
-            runtime.resource_controller.track_grant(message.grant)?;
-        }
-        runtime
-            .execution_outcomes
-            .push(ExecutionOutcome::User(application_id, outcome));
         Ok((runtime.execution_outcomes, runtime.resource_controller))
     }
+
+    /// Notifies all loaded applications that execution is finalizing.
+    fn finalize(&mut self, context: FinalizeContext) -> Result<(), ExecutionError> {
+        let applications = self
```

Comment:
Are we supposed to set `self.is_finalizing` to true here?

Reply:
Oops! Yes, thanks! I fixed it and added some more unit tests for it.